.storybook/grid/GridPanel.tsx

```diff
@@ -0,0 +1,115 @@
+import React, { useEffect, useState, useRef } from 'react';
+import { AddonPanel } from '@storybook/components';
+import { useChannel } from '@storybook/api';
+import { styles } from './gridStyles';
+import { INITIAL_IMAGE_EVENT, IMAGE_SELECTED_EVENT } from './withGrid';
+import { STORY_CHANGED } from '@storybook/core-events';
+import { IframePostMessageService } from '@guardian/grid-client';
+import { Reporter } from '@guardian/grid-client/lib/utils';
+
+const GRID_URL =
+    process.env.NODE_ENV === 'development'
+        ? 'https://media.test.dev-gutools.co.uk'
+        : 'https://media.gutools.co.uk';
+
+const isValidMessage = (data) => data?.crop?.data && data?.image?.data;
+
+export const GridPanel = ({ api, active }) => {
+    const emit = useChannel({});
+    const [isModalOpen, setIsModalOpen] = useState(false);
+    const [selectedImage, setSelectedImage] = useState();
+    const activeButton = useRef(null);
+
+    const onMessage = (event) => {
+        if (event.origin !== GRID_URL) {
+            return;
+        }
+
+        const postMessageService = new IframePostMessageService(event, Reporter.default);
+
+        if (!postMessageService.isValid) {
+            return;
+        }
+
+        const cropAssets = event.data.crop.data.assets;
+        const cropUrl = cropAssets[cropAssets.length - 1]?.secureUrl;
+        setSelectedImage(cropUrl);
+        emit(IMAGE_SELECTED_EVENT, cropUrl);
+
+        setIsModalOpen(false);
+    };
+
+    useEffect(() => {
+        if (isModalOpen) {
+            window.addEventListener('message', onMessage, false);
+        } else {
+            window.removeEventListener('message', onMessage, false);
+        }
+
+        return () => window.removeEventListener('message', onMessage, false);
+    }, [isModalOpen]);
+
+    useEffect(() => {
+        // This panel is reused across stories, so reset when the story is changed
+        const onStoryChanged = () => {
+            setIsModalOpen(false);
+            setSelectedImage(undefined);
+        };
+
+        const onInitialImage = (initialImage) => {
+            setSelectedImage(initialImage);
+        };
+
+        const stopListeningToStoryChanged = api.on(STORY_CHANGED, onStoryChanged);
```

Comment:
As a reader it isn't crystal clear to me that this "stops listening". This appears to be adding a handler that does nothing, rather than removing the existing handler? If this works then I guess this is actually mutating the existing handler reference.

No harm if this works, but if this is confusing to others as well perhaps we could think about removing the events rather than changing them to NoOps in this situation?

Reply:
Oh weirdly I'd seen examples where the return value of the `.on` function was a function to de-register the handler, but looking at the docs that's not the case. So I must have been mistaken. I'll fix. I think the more standard `api.off` is the way to go.